### PR TITLE
fix: workload partitioning - added more error handling for retry attempts

### PR DESCRIPTION
**Pull request by eggfoobar**
I'm noticing some continued flakes and errors crop up around this test.

Some of the flakes below seem to be empty array errors, which should not be the case, so we might be eating an error there. Added some logic to print our any `stderr` and also retry if it's empty in a cluster where cpu partitioning is expected.

https://testgrid.k8s.io/redhat-openshift-ocp-release-4.14-informing#periodic-ci-openshift-release-master-nightly-4.14-e2e-aws-ovn-cpu-partitioning&include-filter-by-regex=CPU%20Partitioning

I also noticed some [sippy](https://sippy.dptools.openshift.org/sippy-ng/jobs/4.14/runs?filters=%7B%22items%22%3A%5B%7B%22columnField%22%3A%22failed_test_names%22%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22%5Bsig-node%5D%5Bapigroup%3Aconfig.openshift.io%5D%20CPU%20Partitioning%20node%20validation%20should%20have%20correct%20cpuset%20and%20cpushare%20set%20in%20crio%20containers%20%5BSuite%3Aopenshift%2Fconformance%2Fparallel%5D%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22not%22%3Atrue%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22never-stable%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22not%22%3Atrue%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22aggregated%22%7D%5D%2C%22linkOperator%22%3A%22and%22%7D&sortField=timestamp&sort=desc) errors and flakes around connection issues and odd states such as `machine-config-daemon` not being available, this seems to be related to heterogenous clusters which alter a machine, the timing might be off when the node is coming back up and the machine-config-daemon is not running yet.

### CODE REVIEW

File: test/extended/cpu_partitioning/crio.go [hunk 1]
```diff
@@ -306,10 +306,13 @@ func collectContainerInfo(ctx context.Context, oc *exutil.CLI, node corev1.Node)
...
 	err = retry.OnError(backoff, shouldRetryExec,
 		func() error {
-			out, _, execErr := e2epod.ExecWithOptions(oc.KubeFramework(), execOptions)
+			out, outStdErr, execErr := e2epod.ExecWithOptions(oc.KubeFramework(), execOptions)
 			if execErr != nil {
 				return execErr
 			}
+			if outStdErr != "" {
+				return fmt.Errorf("err execing command %s", outStdErr)
+			}
```
Comment (ffromani) [inline test/extended/cpu_partitioning/crio.go:315]: I'm still unconvinced that having -any- data on stderr means we failed something. At very least this should be bound to the specific command we run (and still we're assuming this never changes) with a code comment.

File: test/extended/cpu_partitioning/crio.go [hunk 2]
```diff
@@ -321,22 +324,29 @@ func collectContainerInfo(ctx context.Context, oc *exutil.CLI, node corev1.Node)
...
+		"error dialing backend: dial tcp", // Some sort of disruption is happening causing the exec to time out, retry.
+		"container not found",             // On occasion there might be situations where the machine-config-daemon isn't running yet.
+		"err execing command",             // Error when running the command in the container
+	}
 	switch {
+	// This typically means that something interrupted the stream of json coming back, so we need to do the query again.
 	case errors.As(err, &parseErr):
 		return true
-	case strings.Contains(err.Error(), "error dialing backend: dial tcp"):
+	case substringSliceContains(substringErrors, err.Error()):
```
Comment (ffromani) [inline test/extended/cpu_partitioning/crio.go:341]: can we use `errors.As` also here? compare errors with known values? matching the error string should really be the last resort.

File: test/extended/cpu_partitioning/crio.go [hunk 3]
```diff
@@ -313,6 +313,9 @@ func collectContainerInfo(ctx context.Context, oc *exutil.CLI, node corev1.Node)
 			if outStdErr != "" {
 				return fmt.Errorf("err execing command %s", outStdErr)
 			}
+			if isClusterCPUPartitioned && len(out) == 0 {
+				return fmt.Errorf("err execing command, script result was empty but expected values")
+			}
```
Comment (ffromani) [inline test/extended/cpu_partitioning/crio.go:318]: we observed this happening in the wild? what was the context?

**Review by ffromani — COMMENTED** [review]
In general I'd appreciate more context giving rationale for the changes. The changes per se look generally reasonnable, but we should aim to minimize the guesswork involved and try hard to present and record context which drove the changes.

**Comment by eggfoobar** [comment]
> In general I'd appreciate more context giving rationale for the changes. The changes per se look generally reasonnable, but we should aim to minimize the guesswork involved and try hard to present and record context which drove the changes.

Ah, apologies, got a bit tunnel vision with this test. I've updated the description with some more context.

File: test/extended/cpu_partitioning/crio.go (hunk 1 above)
Comment (eggfoobar) [reply test/extended/cpu_partitioning/crio.go:315]: I might me misunderstanding then. In our scenario here we're executing this script in the container. Would it not be only errors that come through the stderr in this situation?

File: test/extended/cpu_partitioning/crio.go (hunk 2 above)
Comment (eggfoobar) [reply test/extended/cpu_partitioning/crio.go:341]: I did try and dig through the return types, unfortunately I didn't find a great way to get a guarantee on the error type being thrown there, so the string seemed to be the next sure thing.

File: test/extended/cpu_partitioning/crio.go (hunk 3 above)
Comment (eggfoobar) [reply test/extended/cpu_partitioning/crio.go:318]: Yeah, that was my bad, updated the description with info on that info. In this example job you'll see it flake with a an empty array

https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-release-master-nightly-4.14-e2e-aws-ovn-cpu-partitioning/1685720543094902784

**Review by ffromani — COMMENTED** [review]
/lgtm

thanks for the extra context

**Pull request merged**